src/CairoTTY.h

```diff
@@ -93,21 +94,29 @@ class ICairoTTYProtected
 class ICharPreprocessor
 {
 public:
-    virtual void process(ICairoTTYProtected &ctty, gunichar c) = 0;
+    virtual void process(ICairoTTYProtected &ctty, unsigned char c) = 0;
 
     virtual ~ICharPreprocessor()
     {}
 };
 
+class ICodepageTranslator
+{
+public:
+    virtual bool translate(unsigned char in, gunichar &out) = 0;
```

Comment:
I thought that multi byte characters are not relevant for dotprint. Thats the reason why i implemented it like this. I am not sure whether any line printer preprocessor will make use of those chars, but from my point of view this is used for the very old standard dot matrix printers which even did not support something like that.

I don't like to throw an exception here because this would maybe break the interpretation of an input file. Instead the function now prints out a warning that there was a character which could not be printed. This is usefull when a input file was generated by a program which uses a little different "slang" of the used codec.

Reply:
OK, so we only support single byte encodings. And it returns false when conversion failed. Originally I thought there is no reason to ever return false (as e.g. CP850 or CP852 define all 256 bytes). But e.g. ISO8859-x had a bunch of undefined values... I don't know iISO8859-x is ever relevant for dot matrix printers, but anyway, we can keep the possibility open.

> I don't like to throw an exception here because this would maybe break the interpretation of an input file. Instead the function now prints out a warning that there was a character which could not be printed. This is usefull when a input file was generated by a program which uses a little different "slang" of the used codec.

Different "slang"? If there is an uncovertible byte, then the user selected a wrong encoding, or the file is corrupted, or there is a bug, but certainly the produced output will not be correct. This is, in my opinion dangerous. Also, throwing an exception does not mean the program cannot continue: the exception could be handled by the caller (CairoTTY or even main). This could be combined by some option ("--ignore-invalid-chars") that a user could specify if they really want to try to produce output out of a broken input.

Anyway, feel free to keep ti as it is. I might change this in master.

Reply:
With different slang i mean some special issue in my example file:
Normally the escape command for select quality (0x78) is followed by only one byte (At least this is what i read from the documentation) But with the program i am using there is always another 00 byte after the correct byte.
But you are right. We can implement a similar behaviour with exceptions

Reply:
What you describe does not sound like encoding issue, but like a different escape format - either for different printer, or a different understanding of the spec.

Also, byte 0 is convertible to pretty much anything, and e.g. 0 in CP850 us U+0 in unicode. So the conversion of 0 should succeed and then later in the pipeline CairoTTY could complain about unexpected control char.